lib/Transforms/Functional.cpp

```diff
@@ -0,0 +1,153 @@
+//===- Functional.cpp -------------------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "Transforms/Functional.h"
+#include "mlir/IR/BlockAndValueMapping.h"
+
+using namespace mlir;
+using functional::impl::SequencePattern;
+
+Operation *mlir::functional::findNearestScope(Operation *op) {
```

Comment:
Ah sorry, I get it now that I see the impl.
Do you already see your mechanisms apply to generic isolated form above?
I am also wondering how this relates to the ScopeOp if at all.

Reply:
This predates ScopeOp (which is a special case/utility op) but I removed some of these changes since they aren't needed (for now). I.e. Scope op might be something specific to linalg, whereas the rewrite system in general will work on any op isolated from above (the name is confusing...)